### PR TITLE
Proxy: Support proxy server for SRS. v7.0.16

### PR DESCRIPTION
**Pull request by winlinvip**
Please note that the proxy server is a new architecture or the next version of the Origin Cluster, which allows the publication of multiple streams. The SRS origin cluster consists of a group of origin servers designed to handle a large number of streams.

```text
                         +-----------------------+
                     +---+ SRS Proxy(Deployment) +------+---------------------+
+-----------------+  |   +-----------+-----------+      +                     +
| LB(K8s Service) +--+               +(Redis/MESH)      + SRS Origin Servers  +
+-----------------+  |   +-----------+-----------+      +    (Deployment)     +
                     +---+ SRS Proxy(Deployment) +------+---------------------+
                         +-----------------------+
```

The new origin cluster is designed as a collection of proxy servers. For more information, see [Discussion #3634](https://github.com/ossrs/srs/discussions/3634). If you prefer to use the old origin cluster, please switch to a version before SRS 6.0.

A proxy server can be used for a set of origin servers, which are isolated and dedicated origin servers. The main improvement in the new architecture is to store the state for origin servers in the proxy server, rather than using MESH to communicate between origin servers. With a proxy server, you can deploy origin servers as stateless servers, such as in a Kubernetes (K8s) deployment.

Now that the proxy server is a stateful server, it uses Redis to store the states. For faster development, we use Go to develop the proxy server, instead of C/C++. Therefore, the proxy server itself is also stateless, with all states stored in the Redis server or cluster. This makes the new origin cluster architecture very powerful and robust.

The proxy server is also an architecture designed to solve multiple process bottlenecks. You can run hundreds of SRS origin servers with one proxy server on the same machine. This solution can utilize multi-core machines, such as servers with 128 CPUs. Thus, we can keep SRS single-threaded and very simple. See https://github.com/ossrs/srs/discussions/3665#discussioncomment-6474441 for details.

```text
                                       +--------------------+
                               +-------+ SRS Origin Server  +
                               +       +--------------------+
                               +
+-----------------------+      +       +--------------------+
+ SRS Proxy(Deployment) +------+-------+ SRS Origin Server  +
+-----------------------+      +       +--------------------+
                               +
                               +       +--------------------+
                               +-------+ SRS Origin Server  +
                                       +--------------------+
```

Keep in mind that the proxy server for the Origin Cluster is designed to handle many streams. To address the issue of many viewers, we will enhance the Edge Cluster to support more protocols.

```text
+------------------+                                               +--------------------+
+ SRS Edge Server  +--+                                    +-------+ SRS Origin Server  +
+------------------+  +                                    +       +--------------------+
                      +                                    +
+------------------+  +     +-----------------------+      +       +--------------------+
+ SRS Edge Server  +--+-----+ SRS Proxy(Deployment) +------+-------+ SRS Origin Server  +
+------------------+  +     +-----------------------+      +       +--------------------+
                      +                                    +
+------------------+  +                                    +       +--------------------+
+ SRS Edge Server  +--+                                    +-------+ SRS Origin Server  +
+------------------+                                               +--------------------+
```

With the new Origin Cluster and Edge Cluster, you have a media system capable of supporting a large number of streams and viewers. For example, you can publish 10,000 streams, each with 100,000 viewers.

---------

Co-authored-by: Jacob Su <suzp1984@gmail.com>

### CODE REVIEW

**Comment by winlinvip** [comment]
image

It works pretty well :)

File: trunk/conf/origin1-for-proxy.conf [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+}
+http_api {
+    enabled         on;
+    listen          19851;
+}
+heartbeat {
+    enabled on;
+    interval 3;
+    url http://127.0.0.1:12025/api/v1/srs/register;
+    device_id origin2;
```
Comment (suzp1984) [inline trunk/conf/origin1-for-proxy.conf]: `origin1`?

File: proxy/srs.go [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+	if b, err := v.rdb.Get(ctx, v.redisKeyServers()).Bytes(); err == nil {
+		if err := json.Unmarshal(b, &serverKeys); err != nil {
+			return errors.Wrapf(err, "unmarshal key=%v servers %v", v.redisKeyServers(), string(b))
+		}
+	}
+
+	// Check each server expiration, if not exists in redis, remove from servers.
+	for i, serverKey := range serverKeys {
+		if _, err := v.rdb.Get(ctx, serverKey).Bytes(); err != nil {
+			serverKeys = append(serverKeys[:i], serverKeys[i+1:]...)
```
Comment (suzp1984) [inline proxy/srs.go:364]: remove a slice's element while iterating them by `range` is an error.
ref:
https://vbauerster.github.io/2017/04/removing-items-from-a-slice-while-iterating-in-go/
use:
https://pkg.go.dev/slices#DeleteFunc

File: trunk/conf/origin1-for-proxy.conf (hunk 1 above)
Comment (winlinvip) [reply trunk/conf/origin1-for-proxy.conf]: Fixed

File: proxy/srs.go (hunk 2 above)
Comment (winlinvip) [reply proxy/srs.go:364]: Fixed. Thanks!

File: proxy/srs.go [hunk 3]
```diff
@@ -0,0 +1,553 @@
...
+	return nil
+}
+
+func (v *srsMemoryLoadBalancer) Pick(ctx context.Context, streamURL string) (*SRSServer, error) {
+	// Always proxy to the same server for the same stream URL.
+	if server, ok := v.picked.Load(streamURL); ok {
+		return server, nil
+	}
+
+	// Gather all servers, alive in 60s ago.
```
Comment (suzp1984) [inline proxy/srs.go]: It's gather server within `srsServerAliveDuration = 300s` ago.

Comment (winlinvip) [reply proxy/srs.go]: Fixed.

**Comment by VyVu** [comment]
> image
> It works pretty well :)

What is the hardware configuration of each server (CPU , RAM , bandwidth)?

File: proxy/rtc.go [hunk 4]
```diff
@@ -0,0 +1,515 @@
...
+		return errors.Wrapf(err, "connect backend for %v", v.StreamURL)
+	}
+
+	// Proxy client message to backend.
+	if v.backendUDP == nil {
+		return nil
+	}
+
+	// Proxy all messages from backend to client.
+	go func() {
```
Comment (jurry) [inline proxy/rtc.go:395]: This causes a goroutine leak and consumes the whole available memory very soon (10 GB in my case). I have a fix and can create a PR. Please let me know, maybe you already solved it @winlinvip

Comment (suzp1984) [reply proxy/rtc.go:395]: Please to send your PR.

**Pull request merged**